### PR TITLE
UTF-8 and Gzip for text files

### PR DESCRIPTION
**Pull request by bartvm**
Adds a helper function to open text files which ensures UTF-8 encoding
is used in Python 2, and allows the user to pass files with Gzip
compression (which can be faster as well as more space efficient on
disk).

### CODE REVIEW

File: fuel/utils/formats.py [hunk 1]
```diff
@@ -1,8 +1,35 @@
...
+    encoding : str, optional
+        The encoding to use (see the codecs documentation_ for supported
+        values). Defaults to `utf-8`.
+
+    .. _documentation:
+    https://docs.python.org/3/library/codecs.html#standard-encodings
+
+    """
+    if filename.endswith('.gz.'):
+        zf = gzip.open(filename, mode)
```
Comment (nouiz) [inline fuel/utils/formats.py]: I already tried this gzip and it was very slow to compute. It use a too small buffer to uncompress. It was way faster to uncompress on the command line wigh gunzip. Did you see speed up with this compared to regular file? If so, with which version of Python?

Comment (bartvm) [inline fuel/utils/formats.py]: Using `io.BufferedReader` (6a32dd44bd10eaf87ae7f41778f4942cd5ac339c) speeds things up a lot, it's now only 10% slower than the command line in my test (I'm comparing with `subprocess.Popen(['zcat', fn], stdout=subprocess.PIPE)`). For 4.5 million lines (670MB uncompressed, 270MB compressed) it takes `zcat` 5.5 seconds and `gzip.open` 6.4.

On my desktop the slowdown is about 2x compared to reading the raw text file (takes 2.9 seconds), but I have an SSD and lots of space; I figured that when reading from a network drive and when the text files take up a lot of space it might definitely be worth using `gzip` (keeping in mind that we have monolingual datasets that are 100GB *compressed*).

File: fuel/utils/formats.py [hunk 2]
```diff
@@ -1,8 +1,36 @@
...
+        The mode with which to open the file. Defaults to `r`.
+    encoding : str, optional
+        The encoding to use (see the codecs documentation_ for supported
+        values). Defaults to `utf-8`.
+
+    .. _documentation:
+    https://docs.python.org/3/library/codecs.html#standard-encodings
+
+    """
+    if filename.endswith('.gz.'):
```
Comment (vdumoulin) [inline fuel/utils/formats.py]: Is the trailing period intentional?

**Comment by vdumoulin** [comment]
Other than my very brief inline comment, LGTM.

File: fuel/utils/formats.py (hunk 2 above)
Comment (bartvm) [inline fuel/utils/formats.py]: Nope, thanks.

**Comment by bartvm** [comment]
Okay, I think this is good to go. In summary, `gzip` file handles are only picklable in Python 3 (I tried fixing it in picklable-itertools for Python 2, but it's a nightmare because the buffered stream makes it really hard to tell where you are in the underlying file). Same goes for UTF-8 encoding, which works fine in Python 3 but in Python 2 becomes unpicklable.

**Comment by vdumoulin** [comment]
Wait, would that mean that for Blocks and under Python2 we wouldn't be able to pickle the main loop within an epoch?

**Comment by bartvm** [comment]
Yes, but only if you use `.gz` files or if you pass `encoding='utf-8'`. Neither is serializable in Python 2 without a lot of custom pickling (that I tried to do this morning, but gave up on after 2 hours).

In general though, you can get away with never using UTF-8 in Python 2 as long as you make sure that the dictionary doesn't use UTF-8 either. I was using this within a project where the dictionary was already created elsewhere as UTF-8, so in that case it's useful to have the keyword argument.

In short, nothing changes about the current behaviour, but if you insist on using the new functionality of reading files in UTF-8 or using `gzip` in Python 2 you will lose picklability, yes.

**Comment by vdumoulin** [comment]
As long as it's not breaking anything existing, I'm okay with that. You should document it, though.

**Comment by dwf** [comment]
Yes, please document it, [and please use the phrase "legacy Python" and not "Python 2" to do so](https://carreau.github.io/posts/planning-an-early-death-for-python-2.html). :wink:

**Comment by bartvm** [comment]
It's documented already as part of the `TextFile` docstring, or do you mean that I should put in the `open_` docstring as well?

@dwf Sure.

**Comment by vdumoulin** [comment]
I'd put it in the `open_` docstring as well, just in case it's reused by future library components or user code.

**Comment by bartvm** [comment]
Done and rebased.
